### ISSUE TITLE
streaming hangs when column family is dropped in the middle

### ISSUE THREAD

**Issue by tgrabiec** [scylladb/scylla#979]
The problem can be reproduced by running the following dtest:
```
update_cluster_layout_tests.py:TestUpdateClusterLayout.simple_add_new_node_while_schema_changes_test
```

When analyzing one of the runs, it looks like streaming session from node2 to node4 failed because node4 no longer has the table which is streamed:

```
INFO  [shard 0] stream_session - [Stream #ce3a9d90-e033-11e5-a099-000000000000] Prepare completed with 127.0.0.4. Receiving 0, sending 1
ERROR [shard 0] stream_session - [Stream #ce3a9d90-e033-11e5-a099-000000000000] stream_transfer_task: Fail to send to 127.0.0.4:0: std::runtime_error (Can't find a column family with UUID b8ac9460-e033-11e5-98c8-000000000000)
ERROR [shard 0] stream_session - [Stream #ce3a9d90-e033-11e5-a099-000000000000] Streaming error occurred
INFO  [shard 0] stream_session - [Stream #ce3a9d90-e033-11e5-a099-000000000000] Session with 127.0.0.4 is complete, state=FAILED
INFO  [shard 0] stream_session - [Stream #ce3a9d90-e033-11e5-a099-000000000000] bytes_sent = 103020, bytes_received = 0
WARN  [shard 0] stream_session - [Stream #ce3a9d90-e033-11e5-a099-000000000000] Stream failed, peers={127.0.0.4}
```

The fact that node4 no longer has it is fine and is the condition which the test tries to exercise - that streaming completes while table is dropped in the middle. The problem is that node4 seems to be stuck and eventually the test times out.

The dtest doesn't run into this failure most of the time because streaming completes so fast or dropping is so slow so that the table is usually dropped after all streaming is done and the test doesn't exercise the condition it is supposed to test. To make the failure more likely, increase the amount of data to stream:

```patch
diff --git a/update_cluster_layout_tests.py b/update_cluster_layout_tests.py
index 27ae6a5..9424471 100644
--- a/update_cluster_layout_tests.py
+++ b/update_cluster_layout_tests.py
@@ -427,7 +427,7 @@ class TestUpdateClusterLayout(Tester):
         self.create_ks(session, 'ks', rf)
         self.create_cf(session, 'cf', read_repair=0.0, columns={'c1': 'text', 'c2': 'text'})
 
-        insert_c1c2(session, keys=range(4000), consistency=consistency)
+        insert_c1c2(session, keys=range(400000), consistency=consistency)
 
         event = threading.Event()
```

**Comment by tgrabiec** [comment]
Related issue #976

**Comment by asias** [comment]
In the log below is not due to remote does not the has cf_id, but the local node(the sender of the mutation) can not find the mutation, probably the cf_id is deleted after we make the reader.

```
ERROR [shard 0] stream_session - [Stream #ce3a9d90-e033-11e5-a099-000000000000] stream_transfer_task: Fail to send to 127.0.0.4:0: std::runtime_error (Can't find a column family with UUID b8ac9460-e033-11e5-98c8-000000000000)
```

**Comment by asias** [comment]
Update: Actually, it is the remote peer throws the no_such_column_family exception in the STREAM_MUTATION_DONE handler.

**Comment by asias** [comment]
Patch sent: streaming: Handle cf is deleted when sending STREAM_MUTATION_DONE

**Comment by asias** [comment]
With my patch, now I saw a different issue in some of the runs.

FAIL: simple_add_new_node_while_schema_changes_test (update_cluster_layout_tests.TestUpdateClusterLayout)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/asias/src/cloudius-systems/scylla-dtest/update_cluster_layout_tests.py", line 461, in simple_add_new_node_while_schema_changes_test
    self.assertEqual(len(result), 1000, len(result))
AssertionError: 960
-------------------- >> begin captured logging << --------------------
dtest: DEBUG: cluster ccm directory: /home/asias/.dtest/dtest-cax08q
--------------------- >> end captured logging << ---------------------

----------------------------------------------------------------------

**Comment by slivne** [comment]
@asias I have added your tests - yet have not enabled them in scylla_tests as your patch is not yet merged (and it will close this issue and lets eable them)

Once it is merged - please open another bug for the "new" issue so we can track it.

**Comment by asias** [comment]
To fix the test, simple_add_new_node_while_schema_changes_test, we also need the fixes for #1000

**Comment by asias** [comment]
Patches are sent.
